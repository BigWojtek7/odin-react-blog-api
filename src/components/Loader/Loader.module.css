/* Loader.module.css */
.overlay {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.2);
  z-index: 99999;
}

.loaderContainer {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1rem;
}

.spinner {
  width: 50px;
  height: 50px;
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #3498db;
  animation: spin 1s linear infinite;
}

.text {
  color: #fff;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
