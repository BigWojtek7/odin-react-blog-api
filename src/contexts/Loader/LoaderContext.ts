import { createContext } from 'react';

interface LoaderContextType {
  isLoading: boolean;
  loaderText: string;
  start: (loaderText?: string) => void;
  stop: () => void;
}

const LoaderContext = createContext<LoaderContextType | null>(null);

export default LoaderContext;
