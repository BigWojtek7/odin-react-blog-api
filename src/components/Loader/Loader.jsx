import { useContext, useEffect, useState } from 'react';
import LoaderContext from '../../contexts/Loader/LoaderContext';

import styles from './Loader.module.css';

function Loader() {
  const { isLoading, loaderText } = useContext(LoaderContext);
  const [isDelay, setIsDelay] = useState(false);

  useEffect(() => {
    if (isLoading) {
      const timeoutId = setTimeout(() => {
        setIsDelay(true);
      }, 3000);
      return () => clearTimeout(timeoutId);
    } else {
      setIsDelay(false);
    }
  }, [isLoading]);
  return (
    <>
      {isLoading ? (
        <div className={styles.overlay}>
          <div className={styles.loaderContainer}>
            <span className={styles.spinner} />
            <span className={styles.text}>{loaderText}</span>
            {isDelay && (
              <span className={styles.text}>
                loading may take longer when server wakes up
              </span>
            )}
          </div>
        </div>
      ) : null}
    </>
  );
}

export default Loader;
