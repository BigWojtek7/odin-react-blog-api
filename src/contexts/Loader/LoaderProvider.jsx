import { useCallback, useEffect, useState } from 'react';

import LoaderContext from './LoaderContext';

const LoaderProvider = ({ children }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [loaderText, setLoaderText] = useState('');
  const [loaderStack, setLoaderStack] = useState([]);

  const start = useCallback((loaderText = 'Loading...') => {
    setLoaderText(loaderText);
    setLoaderStack((prevStack) => [...prevStack, true]);
  }, []);

  const stop = useCallback(() => {
    setLoaderStack((prevStack) => prevStack.slice(1));
  }, []);

  // useEffect(() => {
  //   console.log('Loader stack changed:', loaderStack);
  //   if (!loaderStack.length) {
  //     setIsLoading(false);
  //     return;
  //   }
  //   setIsLoading(true);
  // }, [loaderStack]);

  useEffect(() => {
    // console.log('Loader stack changed:', loaderStack);
    setIsLoading(loaderStack.length > 0);
  }, [loaderStack]);

  return (
    <LoaderContext.Provider value={{ isLoading, loaderText, start, stop }}>
      {children}
    </LoaderContext.Provider>
  );
};

export default LoaderProvider;
