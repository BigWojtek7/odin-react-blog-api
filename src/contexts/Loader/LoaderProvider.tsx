import { useCallback, useEffect, useState } from 'react';

import LoaderContext from './LoaderContext';

const LoaderProvider = ({ children }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [loaderText, setLoaderText] = useState('');
  const [loaderStack, setLoaderStack] = useState([]);

  const start = useCallback((loaderText = 'Loading...') => {
    setLoaderText(loaderText);
    setLoaderStack((prevStack) => [...prevStack, true]);
  }, []);

  const stop = useCallback(() => {
    setLoaderStack((prevStack) => prevStack.slice(1));
  }, []);

  useEffect(() => {
    setIsLoading(loaderStack.length > 0);
  }, [loaderStack]);

  return (
    <LoaderContext.Provider value={{ isLoading, loaderText, start, stop }}>
      {children}
    </LoaderContext.Provider>
  );
};

export default LoaderProvider;
