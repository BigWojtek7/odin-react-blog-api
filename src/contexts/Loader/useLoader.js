import LoaderContext from './LoaderContext';

import { useContext } from 'react';

const useLoader = () => {
  const loaderContext = useContext(LoaderContext);

  if (!loaderContext) {
    throw new Error(
      'Please use useLoader inside the context of LoaderProvider'
    );
  }
  return {
    start: loaderContext.start,
    stop: loaderContext.stop,
  };
};

export default useLoader;
