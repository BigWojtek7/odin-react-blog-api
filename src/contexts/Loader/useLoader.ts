import LoaderContext from './LoaderContext';

import { useContext } from 'react';

const useLoader = () => {
  const loaderContext = useContext(LoaderContext);

  if (!loaderContext) {
    throw new Error(
      'Please use useLoader inside the context of LoaderProvider'
    );
  }
  return {
    loaderText: loaderContext.loaderText,
    isLoading: loaderContext.isLoading,
    start: loaderContext.start,
    stop: loaderContext.stop,
  };
};

export default useLoader;
