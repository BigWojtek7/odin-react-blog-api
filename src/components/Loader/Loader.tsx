import { useEffect, useState } from 'react';

import styles from './Loader.module.css';
import useLoader from '../../contexts/Loader/useLoader';

function Loader() {
  const { isLoading, loaderText } = useLoader();
  const [isDelay, setIsDelay] = useState(false);

  console.log(isLoading)
  useEffect(() => {
    if (isLoading) {
      const timeoutId = setTimeout(() => {
        setIsDelay(true);
      }, 3000);
      return () => clearTimeout(timeoutId);
    } else {
      setIsDelay(false);
    }
  }, [isLoading]);
  return (
    <>
      {isLoading ? (
        <div className={styles.overlay}>
          <div className={styles.loaderContainer}>
            <span className={styles.spinner} />
            <span className={styles.text}>{loaderText}</span>
            {isDelay && (
              <span className={styles.text}>
                loading may take longer when server wakes up
              </span>
            )}
          </div>
        </div>
      ) : null}
    </>
  );
}

export default Loader;
