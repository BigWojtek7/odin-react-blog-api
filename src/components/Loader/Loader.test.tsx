import { render } from '@testing-library/react';
import Loader from './Loader';

import LoaderContext from '../../contexts/Loader/LoaderContext';

describe('Loader component tests', () => {
  const loaderContextValue = {
    isLoading: true,
    loaderText: 'Loading...',
    start: () => {},
    stop: () => {},
  };

  it('renders Loader component with default context', () => {
    const { container } = render(
      <LoaderContext.Provider value={loaderContextValue}>
        <Loader />
      </LoaderContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('displays loading text when isLoading is true', () => {
    const { getByText } = render(
      <LoaderContext.Provider value={loaderContextValue}>
        <Loader />
      </LoaderContext.Provider>
    );

    expect(getByText('Loading...')).toBeInTheDocument();
  });
});
