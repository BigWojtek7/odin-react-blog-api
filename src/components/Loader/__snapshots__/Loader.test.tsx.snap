// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loader component tests > renders Loader component with default context 1`] = `
<div>
  <div
    class="_overlay_bf668d"
  >
    <div
      class="_loaderContainer_bf668d"
    >
      <span
        class="_spinner_bf668d"
      />
      <span
        class="_text_bf668d"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
